jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java

```diff
@@ -24,61 +24,84 @@
-  private final URI collectorUri;
+  /**
+   * @param endpoint Jaeger REST endpoint consuming jaeger.thrift, e.g
+   * http://localhost:14268/api/traces
+   */
+  public HttpSender(String endpoint) {
+    this(endpoint, ONE_MB_IN_BYTES);
+  }
+
+  /**
+   * @param endpoint Jaeger REST endpoint consuming jaeger.thrift, e.g
+   * http://localhost:14268/api/traces
+   * @param maxPayloadBytes max bytes to serialize as payload
+   */
+  public HttpSender(String endpoint, int maxPayloadBytes) {
+    this(HttpUrl.parse(String.format("%s?%s", endpoint, HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM)),
+        maxPayloadBytes);
+  }
 
   /**
-   * @param endpoint Jaeger REST endpoint consuming jaeger.thrift, e.g http://localhost:14268/api/traces
-   * @param maxPacketSize max packet size
+   * @param url Jaeger url accepting jaeger.thrift. This needs to include query params, etc.
+   * @param maxPayloadBytes max bytes to serialize as payload
    */
-  public HttpSender(String endpoint, int maxPacketSize) {
-    super(PROTOCOL_FACTORY, maxPacketSize);
-    this.collectorUri = constructCollectorUri(endpoint);
+  public HttpSender(HttpUrl url, int maxPayloadBytes) {
+    super(PROTOCOL_FACTORY, maxPayloadBytes);
+    if (url == null) {
+      throw new IllegalArgumentException("Could not parse url.");
+    }
+    this.collectorUrl = url;
   }
 
   @Override
-  public void send(Process process, List<com.uber.jaeger.thriftjava.Span> spans) throws TException {
+  public void send(Process process, List<Span> spans) throws TException {
     Batch batch = new Batch(process, spans);
     byte[] bytes = SERIALIZER.serialize(batch);
 
+    RequestBody body = RequestBody.create(MediaType.parse("application/x-thrift"), bytes);
+    Request request = new Request.Builder().url(collectorUrl).post(body).build();
+    Response response;
     try {
-      HttpPost httpPost = new HttpPost(this.collectorUri);
-      httpPost.setEntity(new ByteArrayEntity(bytes));
-      HttpResponse response = httpClient.execute(httpPost);
-      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
-        throw new TException("Could not send " + spans.size() + " spans, response "
-            + response.getStatusLine().getStatusCode());
-      }
+      response = httpClient.newCall(request).execute();
     } catch (IOException e) {
-      throw new TException("Could not send " + spans.size() + ", spans", e);
+      throw new TException("Could not make request to send " + spans.size() + ", spans", e);
     }
-  }
 
-  private URI constructCollectorUri(String endpoint) {
-    try {
-      return new URI(String.format("%s?%s", endpoint, HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM));
-    } catch (URISyntaxException e) {
-      throw new IllegalArgumentException("Wrong collector host", e);
+    if (!response.isSuccessful()) {
+      String responseBody;
+      try {
+        responseBody = response.body() != null ? response.body().string() : "null";
```

Comment:
I could, but it would not protect us from `NPE`.
Specifically, `response.body()` itself can be `null`, so the `.string()` access can throw a `NPE`. Also, `body.string()` consumes the response and is different from `body.toString()`.